src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -333,8 +336,69 @@ public String executor() {
         }
 
         @Override
-        public void messageReceived(ShardActiveRequest request, TransportChannel channel) throws Exception {
-            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+        public void messageReceived(final ShardActiveRequest request, final TransportChannel channel) throws Exception {
+            // make sure shard is really there before register cluster state obsever
+            ShardId shardId = request.shardId;
+            IndexService indexService = indicesService.indexService(shardId.index().getName());
+            if (indexService != null && indexService.indexUUID().equals(request.indexUUID)) {
+                if (!indexService.hasShard(shardId.id())) {
+                    channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+                }
+            } else {
+                channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+            }
+            // create observer just in case
+            ClusterStateObserver observer = new ClusterStateObserver(clusterService, request.timeout, logger);
+            // check if shard is active. if so, all is good
+            boolean shardActive = shardActive(request);
+            if (shardActive) {
+                channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+            } else {
+                // shard is not active, might be POST_RECOVERY so check if cluster state changed inbetween or wait for next change
+                observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
```

Comment:
the exception must be the first element after the log message then you can drop the second log line